Hera gnu build broken

The `gsi_hera.gnu.lua` from `develop` at 27c03e83 is missing `-rocky8` on the the `prepend_path` line.  Even after fixing this in a working copy of `develop`
```
-prepend_path("MODULEPATH", "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
```
the gnu build fails with the message
```
[  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/jp_pkind.f90.o
cd /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/build/src/mgbf && /apps/gnu/gcc-9.2.0/bin/gfortran -DNDEBUG -I/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/build/src/mgbf/include/mgbf -g -fbacktrace -fconvert=big-endian -O3 -DNDEBUG -O3 -Jinclude/mgbf -c /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind.f90 -o CMakeFiles/mgbf.dir/jp_pkind.f90.o
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind.f90:22:4:

   22 | use mpi
      |    1
Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
compilation terminated.
make[2]: *** [src/mgbf/CMakeFiles/mgbf.dir/build.make:88: src/mgbf/CMakeFiles/mgbf.dir/jp_pkind.f90.o] Error 1
```

This issue is opened to document the fact that the Hera gnu build is broken.   This issue will remain open until the Hera gnu build is fixed or the decision is made to **_not_** support a GSI gnu build on Hera.

Based on [comments](https://github.com/NOAA-EMC/GSI-utils/issues/66#issuecomment-2627296301) from @RatkoVasic-NOAA in GSI-utils [#66](https://github.com/NOAA-EMC/GSI-utils/issues/66), I made the following changes to a working copy of develop `modulefiles/gsi_hera.gnu.lua`

```
-prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/scratch4/NCEPDEV/stmp/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulef
iles/Core")
 --Needed for openmpi build
-prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
+--prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")

 local python_ver=os.getenv("python_ver") or "3.11.6"
 local stack_gnu_ver=os.getenv("stack_gnu_ver") or "9.2.0"
-local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.5"
+local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.6"
```

The first GSI build attempt failed with
```
[  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/jp_pkind2.f90.o
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind2.f90:22:4:

   22 | use mpi
      |    1
Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind.f90:22:4:

   22 | use mpi
      |    1
Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
compilation terminated.
```

As a test I turned off the MGBF build and tried again. The build got further but failed again
```
 79%] Building Fortran object src/gsi/CMakeFiles/gsi_fortran_obj.dir/cplr_regional_io.f90.o
[ 80%] Building Fortran object src/gsi/CMakeFiles/gsi_fortran_obj.dir/hybrid_ensemble_isotropic.F90.o
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/gsi/hybrid_ensemble_isotropic.F90:107:6:

  107 |   use mg_intstate
      |      1
Fatal Error: Cannot open module file 'mg_intstate.mod' for reading at (1): No such file or directory
compilation terminated.
make[2]: *** [src/gsi/CMakeFiles/gsi_fortran_obj.dir/build.make:2610: src/gsi/CMakeFiles/gsi_fortran_obj.dir/hybrid_ensemble_isotropic.F90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:203: src/gsi/CMakeFiles/gsi_fortran_obj.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

Line 107 of src/gsi/hybrid_ensemble_isotropic.F90 is in the block
```
! For MGBF
  use mg_intstate
  use mg_timers
```

We do not use MGBF in any of our regional or global ctests. We do not use MGBF in the operational GDAS. Nonetheless, it seems we must compile MGBF in order to build gsi.x.

My guess is that the gnu build will work once the gnu MGBF issues are resolved.

Tagging @TingLei-NOAA and @ShunLiu-NOAA to confirm that we must compile MGBF in order to build gsi.x and enkf.x. If true we need to

1. fix MGBF so that it compiles with gnu, or
2. modify the GSI build so that MGBF is an optional component

Russ
I will check why it is a problem for gnu as soon as possible.
For mgbf in gsi that issue/pr was created because rtma is planning to use
it
Ting

______________________________

Ting Lei

Physical Scientist, Contractor with Lynker in support of

EMC/NCEP/NWS/NOAA

5830 University Research Ct., Cubicle 2765

College Park, MD 20740

***@***.***

301-683-3624


On Fri, Jan 31, 2025 at 9:16 AM RussTreadon-NOAA ***@***.***>
wrote:

> Based on comments
> <https://github.com/NOAA-EMC/GSI-utils/issues/66#issuecomment-2627296301>
> from @RatkoVasic-NOAA <https://github.com/RatkoVasic-NOAA> in GSI-utils
> #66 <https://github.com/NOAA-EMC/GSI-utils/issues/66>, I made the
> following changes to a working copy of develop
> modulefiles/gsi_hera.gnu.lua
>
> -prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
> +prepend_path("MODULEPATH", "/scratch4/NCEPDEV/stmp/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulef
> iles/Core")
>  --Needed for openmpi build
> -prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
> +--prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
>
>  local python_ver=os.getenv("python_ver") or "3.11.6"
>  local stack_gnu_ver=os.getenv("stack_gnu_ver") or "9.2.0"
> -local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.5"
> +local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.6"
>
> The first GSI build attempt failed with
>
> [  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/jp_pkind2.f90.o
> /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind2.f90:22:4:
>
>    22 | use mpi
>       |    1
> Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
> /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind.f90:22:4:
>
>    22 | use mpi
>       |    1
> Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
> compilation terminated.
>
> As a test I turned off the MGBF build and tried again. The build got
> further but failed again
>
>  79%] Building Fortran object src/gsi/CMakeFiles/gsi_fortran_obj.dir/cplr_regional_io.f90.o
> [ 80%] Building Fortran object src/gsi/CMakeFiles/gsi_fortran_obj.dir/hybrid_ensemble_isotropic.F90.o
> /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/gsi/hybrid_ensemble_isotropic.F90:107:6:
>
>   107 |   use mg_intstate
>       |      1
> Fatal Error: Cannot open module file 'mg_intstate.mod' for reading at (1): No such file or directory
> compilation terminated.
> make[2]: *** [src/gsi/CMakeFiles/gsi_fortran_obj.dir/build.make:2610: src/gsi/CMakeFiles/gsi_fortran_obj.dir/hybrid_ensemble_isotropic.F90.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:203: src/gsi/CMakeFiles/gsi_fortran_obj.dir/all] Error 2
> make: *** [Makefile:146: all] Error 2
>
> Line 107 of src/gsi/hybrid_ensemble_isotropic.F90 is in the block
>
> ! For MGBF
>   use mg_intstate
>   use mg_timers
>
> We do not use MGBF in any of our regional or global ctests. We do not use
> MGBF in the operational GDAS. Nonetheless, it seems we must compile MGBF in
> order to build gsi.x.
>
> My guess is that the gnu build will work once the gnu MGBF issues are
> resolved.
>
> Tagging @TingLei-NOAA <https://github.com/TingLei-NOAA> and @ShunLiu-NOAA
> <https://github.com/ShunLiu-NOAA> to confirm that we must compile MGBF in
> order to build gsi.x and enkf.x. If true we need to
>
>    1. fix MGBF so that it compiles with gnu, or
>    2. modify the GSI build so that MGBF is an optional component
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/GSI/issues/831#issuecomment-2627457468>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APEFS7HQHRIXQFQJTTRN5ZL2NOAR7AVCNFSM6AAAAABWFTIBD2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMRXGQ2TONBWHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

@TingLei-NOAA , did you make the following changes in your working copy of `modulefiles/gsi_hera.gnu.lua`
```
--- a/modulefiles/gsi_hera.gnu.lua
+++ b/modulefiles/gsi_hera.gnu.lua
@@ -1,13 +1,13 @@
 help([[
 ]])

-prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/scratch4/NCEPDEV/stmp/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
 --Needed for openmpi build
-prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
+--prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")

 local python_ver=os.getenv("python_ver") or "3.11.6"
 local stack_gnu_ver=os.getenv("stack_gnu_ver") or "9.2.0"
-local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.5"
+local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.6"
 local cmake_ver=os.getenv("cmake_ver") or "3.23.1"
 local prod_util_ver=os.getenv("prod_util_ver") or "2.1.1"
 local openblas_ver=os.getenv("openblas_ver") or "0.3.24"
```
I made the above changes to a working copy of GSI `develop`.   Look in Hera `/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop`.  After this I executed
```
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop$ cd ush
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/ush$ export COMPILER=gnu
Hera(hfe07):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/ush$ ./build.sh > build.log 2>&1
```
`tail -25 build.log` returns
```
Dependee "/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/build/src/mgbf/CMakeFiles/CMakeDirectoryInformation.cmake" is newer than depender "/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/build/src/mgbf/CMakeFiles/mgbf.dir/depend.internal".
Scanning dependencies of target mgbf
[  0%] Building C object src/gsi/CMakeFiles/gsi_c_obj.dir/blockIO.c.o
[  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/jp_pkind2.f90.o
[  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/jp_pkind.f90.o
[  0%] Building Fortran object src/mgbf/CMakeFiles/mgbf.dir/kinds.f90.o
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind.f90:22:4:

   22 | use mpi
      |    1
Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/develop/src/mgbf/jp_pkind2.f90:22:4:

   22 | use mpi
      |    1
Fatal Error: Cannot open module file 'mpi.mod' for reading at (1): No such file or directory
compilation terminated.
compilation terminated.
make[2]: *** [src/mgbf/CMakeFiles/mgbf.dir/build.make:88: src/mgbf/CMakeFiles/mgbf.dir/jp_pkind.f90.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [src/mgbf/CMakeFiles/mgbf.dir/build.make:101: src/mgbf/CMakeFiles/mgbf.dir/jp_pkind2.f90.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:195: src/mgbf/CMakeFiles/mgbf.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  0%] Built target gsi_c_obj
make: *** [Makefile:146: all] Error 2
```

--- a/modulefiles/gsi_hera.gnu.lua
+++ b/modulefiles/gsi_hera.gnu.lua
@@ -2,8 +2,6 @@ help([[
 ]])
 
 prepend_path("MODULEPATH", "/scratch4/NCEPDEV/stmp/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
---Needed for openmpi build
---prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
 
 local python_ver=os.getenv("python_ver") or "3.11.6"
 local stack_gnu_ver=os.getenv("stack_gnu_ver") or "9.2.0"

--- a/modulefiles/gsi_hera.gnu.lua
+++ b/modulefiles/gsi_hera.gnu.lua
@@ -1,13 +1,13 @@
 help([[
 ]])
 
-prepend_path("MODULEPATH", "/contrib/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
+prepend_path("MODULEPATH", "/scratch4/NCEPDEV/stmp/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/Core")
 --Needed for openmpi build
-prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
+--prepend_path("MODULEPATH", "/scratch1/NCEPDEV/jcsda/jedipara/spack-stack/modulefiles")
 
 local python_ver=os.getenv("python_ver") or "3.11.6"
 local stack_gnu_ver=os.getenv("stack_gnu_ver") or "9.2.0"
-local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.5"
+local stack_openmpi_ver=os.getenv("stack_openmpi_ver") or "4.1.6"
 local cmake_ver=os.getenv("cmake_ver") or "3.23.1"
 local prod_util_ver=os.getenv("prod_util_ver") or "2.1.1"
 local openblas_ver=os.getenv("openblas_ver") or "0.3.24"

--- a/src/mgbf/CMakeLists.txt
+++ b/src/mgbf/CMakeLists.txt
@@ -30,6 +30,8 @@ endif()
 if(NOT CMAKE_BUILD_TYPE MATCHES "Debug")
   add_definitions(-DNDEBUG)
 endif()
+# Find MPI Package
+find_package(MPI REQUIRED)
 
 list(APPEND MGBF_SRC
 kinds.f90
@@ -63,6 +65,8 @@ add_library(${PROJECT_NAME}::mgbf ALIAS mgbf)
 set_target_properties(mgbf PROPERTIES Fortran_MODULE_DIRECTORY "${module_dir}")
 target_include_directories(mgbf PUBLIC $<BUILD_INTERFACE:${module_dir}>
                                            $<INSTALL_INTERFACE:include/mgbf>)
+target_link_libraries(mgbf  PUBLIC MPI::MPI_Fortran)
+
 
 install(DIRECTORY ${module_dir} DESTINATION ${CMAKE_INSTALL_PREFIX}/include)
 

